src/v/cluster/rm_stm.cc

```diff
@@ -792,11 +792,9 @@ ss::future<result<raft::replicate_result>> rm_stm::replicate(
 
 ss::future<std::error_code>
 rm_stm::transfer_leadership(std::optional<model::node_id> target) {
-    return _state_lock.hold_write_lock().then(
-      [this, target](ss::basic_rwlock<>::holder unit) {
-          return _c->do_transfer_leadership(target).finally(
-            [u = std::move(unit)] {});
-      });
+    auto units = co_await _state_lock.hold_write_lock();
+    co_await checkpoint_in_memory_state();
```

Comment:
if the checkout or do_transfer_leadership fails we need explicitly step down to avoid writing requests after the checkpoint